src/main/protobuf/remote_output_service.proto

```diff
@@ -0,0 +1,314 @@
+  // when loading objects from the Content Addressable Storage.
+  string instance_name = 3;
+
+  // The digest function that the client uses when communicating with
+  // the remote execution system. The remote output service uses this
+  // value to ensure that FileStatus responses contain digests that were
+  // computed with right digest function.
+  build.bazel.remote.execution.v2.DigestFunction.Value digest_function = 4;
+
+  // The absolute path at which the remote output service exposes its
+  // output paths, as seen from the perspective of the client.
+  //
+  // This value needs to be provided by the client, because file system
+  // namespace virtualization may cause this directory to appear at a
+  // location that differs from the one used by the service.
+  //
+  // The purpose of this field is to ensure that the remote output
+  // service is capable of expanding symbolic links containing absolute
+  // paths.
+  string output_path_prefix = 5;
+
+  // A map of paths on the system that will become symbolic links
+  // pointing to locations inside the output path. Similar to
+  // output_path_prefix, this option is used to ensure the remote output
+  // service is capable of expanding symbolic links.
+  //
+  // Map keys are absolute paths, while map values are paths that are
+  // relative to the output path.
+  map<string, string> output_path_aliases = 6;
+}
+
+message InitialOutputPathContents {
+  // The identifier of a previously finalized build whose results are
+  // stored in the output path.
+  string build_id = 1;
+
+  // Paths that have been modified or removed since the build finalized.
+  //
+  // If the remote output service freezes the contents of the output
+  // path between builds, this field can be left empty.
+  repeated string modified_paths = 2;
+}
+
+message StartBuildResponse {
+  // If set, the contents of the output path are almost entirely
+  // identical on the results of a previous build. This information may
+  // be used by the client to prevent unnecessary scanning of the file
+  // system.
+  //
+  // Servers can leave this field unset in case the contents of the
+  // output path are empty, not based on a previous build, if no
+  // tracking of this information is performed, or if the number of
+  // changes made to the output path is too large to be expressed.
+  InitialOutputPathContents initial_output_path_contents = 1;
+
+  // A relative path that the client must append to
+  // StartBuildRequest.output_path_prefix to obtain the full path at
+  // which outputs of the build are stored.
+  //
+  // If the remote output service is incapable of storing the output of
+  // multiple builds, this string may be left empty.
+  string output_path_suffix = 2;
+}
+
+message BatchCreateRequest {
+  // The identifier of the build. The remote output service uses this to
+  // determine which output path needs to be modified.
+  string build_id = 1;
```

Comment:
Is that indirection to get `output_base_id`?

Reply:
Yes. The build ID implies the output base ID. But a single output base ID may be associated with multiple build directories (due to snapshotting/copy-on-write semantics, if provided).

Reply:
This is really a question, not an ask to change it, but why would we not use `StartBuildResponse::output_path_suffix` here (that can also mean as part of the paths) instead?

I guess what I am really going after is that we make Bazel aware of these directories to later use the IDs anyway--I am wondering if we could possibly remove any of those concepts from the API.

Reply:
This is done for a somewhat defensive reason: it allows bb_clientd to detect accidental concurrent reuse of the same `output_path_suffix`. For example, in an environment where bb_clientd is run once, but multiple copies of Bazel are launched in separate containers, it would be harmful if each of those copies tried to start a build with the same `output_path_suffix`.

By letting all `BatchCreate()`, `BatchStat()` and `FinalizeBuild()` calls take the build ID, bb_clientd is capable of detecting this misuse. It can ensure that at most one build is permitted per `output_path_suffix`, rejecting calls that provide the wrong value.

Reply:
Is the `path_prefix` kind of a per-blaze workspace? That could be a better way to name that maybe. Also, some example of all of these paths would be likely very welcome.